stan/math/rev/functor/reduce_sum.hpp

```diff
@@ -20,92 +20,91 @@ template <typename ReduceFunction, typename ReturnType, typename M,
           typename... Args>
 struct reduce_sum_impl<ReduceFunction, require_var_t<ReturnType>, ReturnType, M,
                        Args...> {
+  template <typename T>
+  static T& deep_copy(T& arg) {
```

Comment:
I moved these into reduce_sum_impl cause reduce_sum_impl::operator() uses them now. They're static so that functions in recursive_reducer can use them. They should probably be in namespace internal.

Reply:
and maybe marked `inline`?

Reply:
^ pretty much should mark it all inline

Reply:
btw, all the logic which we created here to read of adjoints from tuple arguments would make up for a great refactor of the ODE code... having ODE RHS functors which are allowed to have any function signature would be nice. Not as badly needed as for `reduce_sum`, but for another day that's a good next project.

Reply:
> would make up for a great refactor of the ODE code

Holy crap that's a good point. Yeah we should definitely do something like this.

Reply:
^operands_and_partials with the tuples as well right?